### PR TITLE
Fix drawing of markup with & in generic wxDVC

### PR DESCRIPTION
**Pull request by vslavik**
@vadz this comes from the wxDVC changes to allow markup in items. See https://github.com/vslavik/poedit/issues/328 for what is happening: markup containing "&amp;&amp;" is rendered as "&&&" and not "&&" as it should.

The cause is obvious: `wxMarkupText::SetMarkupText()` escapes mnemonics (presumably for compatibility with previous uses where this class for labels only). `RenderItemText()` added at the same time *doesn't* expect mnemonics to be there (because it's used for item text where they make no sense).

This patch is the trivial fix for it, but doing this is incredibly stupid: first escape mnemonics, then unescape them.

Wouldn't it be a better fix to

1. Store the text as dumb in `wxMarkupText` and let it be interpreted when *rendered*, not store.
2. Remove `SetMarkupText` because it makes no sense after doing 1. above.
3. Escape mnemonics (previously done when setting text) in `wxMarkupText::Render()`. 

?

I think it would be, but didn't want to do that in case I'm missing something... At first look, it seems all affected code was added for wxDVC's benefit and `wxMarkupText` is private, so should be safe, but...

### CODE REVIEW

**Review by vadz — COMMENTED** [review]
I think this change is not quite right as it makes `RenderItemText()` inconsistent with `Measure()`, which still doesn't call `RemoveMnemonics()`. But, of course, calling it in `Measure()` would make it inconsistent with `Render()`, which would be even worse.

But then I don't really understand where does the problem come from neither and I can't reproduce it with
```
diff --git a/samples/dataview/mymodels.cpp b/samples/dataview/mymodels.cpp
index 1ee62f8..330cc32 100644
--- a/samples/dataview/mymodels.cpp
+++ b/samples/dataview/mymodels.cpp
@@ -434,8 +434,9 @@ void MyListModel::GetValueByRow( wxVariant &variant,

         case Col_TextWithAttr:
             {
-                static const char *labels[5] =
+                static const char *labels[6] =
                 {
+                    "&& and || or",
                     // These strings will look wrong without wxUSE_MARKUP, but
                     // it's just a sample, so we don't care.
                     "<span color=\"#87ceeb\">light</span> and "
```

What am I missing?

**Comment by vslavik** [comment]
> I think this change is not quite right

And I said as much myself, I was hoping you'd come with a better fix yourself. Never mind my confused babbling about how to fix it above, which you wisely ignored, but I *do* thing `wxMarkupText` is broken by design and should be split into two: `wxItemMarkupText` (or my preferred naming: just `wxMarkupText`) and `wxLabelMarkupText` — the latter being current `wxMarkupText` (from *before* your wxDVC-related changes) with accels handling, the former being for plain text with no mnemonics code at all and so suitable for wxDVC. (Doing that also requires a non-insane `wxMarkupParser` mode that doesn't magically escape `&`s.)

> inconsistent with Measure(), which still doesn't call RemoveMnemonics()

It does, actually, in `wxMarkupParserMeasureOutput::OnText()`.

…And so does `wxMarkupParserRenderItemOutput::OnText()` which must be why you're confused about this PR's existence. The latter did it wrongly (see 889b406b0ef25d6b6a0d3138180c3b038594a3a8), but that's not important here.

> But then I don't really understand where does the problem come from neither and I can't reproduce it 

You passed in invalid markup, `&&` instead of properly escaped `&amp;&amp;` (which in my testing shows *empty* item, not `&&` — but maybe I forgot to undo some bit somewhere). Use valid markup and you'll see it:

1. `SetMarkupText` escapes it and assigns `m_text` = `&&amp;&&amp;`
2. `wxMarkupParser` sees the first `&`, sees it as not an entity, leaves it alone.
3. Then it sees `&amp;`, parses it *and doubles it* (this is the insane part), so we now have `&&&` in the output buffer.
4. Ditto for the second occurrence.
4. The result, `&&&&&&` is passed to `OnText`.
5. Where mnenomics are once again removed, resulting in `&&&`.

I'll do the proper fix a bit later in the way described at the top of this comment, unless you object or want to fix it yourself.

**Comment by vslavik** [comment]
Updated as described in the above comment.

**Comment by vadz** [comment]
Sorry, I hadn't understood what did you mean to do before, but now that I see it, I really can't agree with this fix.

I think you've identified the real problem yourself: it's in the step (3) above. I.e. `wxMarkupParser` just shouldn't handle `&amp;` differently from `&` in the first place, it could be justified, with some difficulty, as long as markup was only used for the control labels, but it just can't work for arbitrary strings. And if we fix this, the rest becomes much simpler, see my proposed changes in #419.

I do agree that interpreting the string differently in different methods of the class, as done in #419, is not very elegant, but it's pretty simple, doesn't duplicate any code as this PR and avoids translating the strings, so I think it's much better on balance. Thoughts?

**Comment by vslavik** [comment]
> I really can't agree with this fix.

Specific, articulated critique would be more useful than such a hand-wavy dismissal. If you object to `ExpectsMnemonics()`, I do too now that I know you don't mind changing the behavior even though it may break some user code (in theory, I doubt that it would in practice). If you object to splitting the class, I explained the reasons (in short: correctness and better API than the alternatives) in #419.

> I think you've identified the real problem yourself: it's in the step (3) above

No, not really, that of course doesn't help, but the *real* problem is trying to fit two different kinds of input into a single class. Case in point: see my review of #419, you *have* to handle both cases differently.

> doesn't duplicate any code as this PR

This PR duplicates a *minimal amount* of code and does it only because de-duplicating it would mean larger and noticeably more complicated code.

**Comment by vadz** [comment]
> > I really can't agree with this fix.
>
> Specific, articulated critique would be more useful than such a hand-wavy dismissal.

Sorry, this wasn't meant to be a hand-wavy dismissal in any way, I was just trying to explain why I hadn't objected to this before (answer: because I didn't fully realize what you were proposing) and I did, of course, follow with the reasons I didn't agree (more complicated, some code duplication).

Anyhow, to get back on track, thanks for noticing the problem with `Measure()` (I'll keep all the discussion here rather than split it between #419 and here), you're totally right, I did break it. And to fix it we would indeed need to either add some flag to the existing method or add some new `MeasureItemText()` which doesn't seem appealing at all, especially because we'd still use some flag internally as we won't want to duplicate `wxMarkupParserMeasureOutput`.

Considering this, the choice between my solution and yours becomes less obvious, as mine becomes more complex. I'm not sure if I prefer adding a flag to `Measure()` or splitting the class in two as you advocate, but in any case I think both are acceptable, while you seem to be thinking that only the latter is, so let's go with your approach. I'd still want to change `wxMarkupParser` behaviour, for the reasons explained before, so I'd still like to not apply this PR in the current state but rather merge it with the other one. I can do if you'd like, but probably not before the week-end.

Please let me know if this is OK with you.

**Comment by vslavik** [comment]
> MeasureItemText() which doesn't seem appealing at all, especially because we'd still use some flag internally as we won't want to duplicate `wxMarkupParserMeasureOutput`.

It doesn't appeal to me either, but the only difference between the two Measure* methods could be that one would strip mnemonics before using `wxMarkupParserMeasureOutput`.

> as mine becomes more complex. I'm not sure if I prefer adding a flag to Measure() or splitting the class in two as you advocate, but in any case I think both are acceptable, while you seem to be thinking that only the latter is

I wouldn't go that far, but I think it's *cleaner* and I think I can articulate why: encapsulation. You should be able to use a `wxMarkupText` instance without knowing details about how it was created, but you can't do that with a flag to `Measure()` (you have to know how to interpret the text in it; `Render()` flags are different in that they describe what the caller wants). Another and IMHO better solution (in that it doesn't need knowledge of how the object was constructed to use it) is using a flag would be to pass a flag to ctor/`SetMarkup()` if that's what you prefer?

For completeness, the code duplication in this PR can be avoided by having a `wxMarkupTextBase` too with only `Measure` and a virtual `GetPlainText` to return mnemonics-less text.

> I can do if you'd like, but probably not before the week-end.

It's fine, I'll do it (for some value of "it", see above).

**Comment by vadz** [comment]
Yes, of course, I agree that encapsulation would be better with the 2 classes, initially I just thought that it wasn't worth the extra effort in this case, but I think you've changed my mind. I also discarded the idea of `wxMarkupTextBase` because `Render()` can't be a virtual method because the current `Render()` and `RenderItemText()` take different parameters, but you're also right that this doesn't prevent us from using it only for `Measure()` which does take the same parameters in both cases.

I also thought about the solution with `bool m_isLabel` field, but I didn't like it because this would require us to do something in `Render()` if `!m_isLabel` and in `RenderItemText()` if `m_isLabel` and even if asserting is not difficult, of course, it doesn't really look like a big win for encapsulation or good design.

*TL;DR*: I think using `wxMarkupTextBase` with just `Measure()` in it and two derived classes, with one of them defining only `Render()` and the other one -- only `RenderItemText()` is the best one.

**Comment by vslavik** [comment]
Apologies for the delay. I now finally force-pushed updated version that is a mix of this and your #419. I think this is how you meant it, but please let me know if you’d like some other changes to it.

**Review by vadz — APPROVED** [review]
I think this is pretty nice, thanks a lot!
